cmd/snap-update-ns/bootstrap.c

```diff
@@ -0,0 +1,166 @@
+    // it should be the snap name.
+    size_t argv0_len = strnlen(buf, buf_size);
+    if (argv0_len + 1 >= num_read) {
+        return NULL;
+    }
+    return &buf[argv0_len + 1];
+}
+
+// setns_into_snap switches mount namespace into that of a given snap.
+static int
+setns_into_snap(const char* snap_name)
+{
+    // Construct the name of the .mnt file to open.
+    char buf[PATH_MAX];
+    int n = snprintf(buf, sizeof buf, "/run/snapd/ns/%s.mnt", snap_name);
+    if (n >= sizeof buf || n < 0) {
+        bootstrap_errno = errno;
+        bootstrap_msg = "cannot format mount namespace file name";
+        return -1;
+    }
+
+    // Open the mount namespace file, note that we don't specify O_NOFOLLOW as
+    // that file is always a special, broken symbolic link.
+    int fd = open(buf, O_RDONLY | O_CLOEXEC);
+    if (fd < 0) {
+        bootstrap_errno = errno;
+        bootstrap_msg = "cannot open mount namespace file";
+        return -1;
+    }
+
+    // Switch to the mount namespace of the given snap.
+    int err = setns(fd, CLONE_NEWNS);
+    if (err < 0) {
+        bootstrap_errno = errno;
+        bootstrap_msg = "cannot switch mount namespace";
+    };
+
+    close(fd);
+    return err;
+}
+
+// partially_validate_snap_name performs partial validation of the given name.
+// The goal is to ensure that there are no / or .. in the name.
+int partially_validate_snap_name(const char* snap_name)
+{
+    // NOTE: neither set bootstrap_{msg,errno} but the return value means that
+    // bootstrap does nothing. The name is re-validated by golang.
+    if (strstr(snap_name, "..") != NULL) {
+        return -1;
+    }
+    if (strchr(snap_name, '/') != NULL) {
+        return -1;
+    }
+}
+
+// bootstrap prepares snap-update-ns to work in the namespace of the snap given
+// on command line.
+void bootstrap(void)
+{
+// NOTE: This lets us use cgo/go to write tests without running the bulk
+// of the code automatically. In snapd we can just set the required
+// environment variable.
+#define TRIGGER_KEY "SNAPD_INTERNAL"
+#define TRIGGER_VAL "x-switch-namespace=1,"
+    const char* snapd_internal = getenv(TRIGGER_KEY);
+    if (snapd_internal == NULL || strstr(snapd_internal, TRIGGER_VAL) == NULL) {
+        bootstrap_errno = 0;
+        bootstrap_msg = "bootstrap is not enabled, set " TRIGGER_KEY "=" TRIGGER_VAL;
+        return;
+    }
+#undef TRIGGER_KEY
+#undef TRIGGER_VAL
+
+    // We don't have argc/argv so let's imitate that by reading cmdline
+    char cmdline[1024];
+    memset(cmdline, 0, sizeof cmdline);
+    ssize_t num_read;
+    if ((num_read = read_cmdline(cmdline, sizeof cmdline)) < 0) {
+        return;
+    }
+
+    // Find the name of the snap by scanning the cmdline.  If there's no snap
+    // name given, just bail out. The go parts will scan this too.
+    const char* snap_name = find_snap_name(cmdline, sizeof cmdline, num_read);
+    if (snap_name == NULL || *snap_name == '\0') {
```

Comment:
The second case would be more safely handled inside find_snap_name. If it got an empty string, it didn't actually find a snap name.